### PR TITLE
Support  debug protocol.

### PR DESCRIPTION
**Pull request by yaohaizh**
@andxu @akaroml @testforstephen

### CODE REVIEW

**Comment by yaohaizh** [comment]
https://github.com/Microsoft/vscode-java-debug/issues/235

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+
+    private void stepin(IDebugAdapterContext context, ThreadReference thread) {
+        StepRequest request = DebugUtility.createStepIntoRequest(thread, context.getStepFilters().classNameFilters);
+        context.getDebugSession().getEventHub().stepEvents().filter(debugEvent -> request.equals(debugEvent.event.request())).take(1).subscribe(debugEvent -> {
+            debugEvent.shouldResume = false;
+            // Have to send to events to keep the UI sync with the step in operations:
+            context.getProtocolServer().sendEvent(new Events.StoppedEvent("step", thread.uniqueID()));
+            context.getProtocolServer().sendEvent(new Events.ContinuedEvent(thread.uniqueID()));
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: why to send continued event to vscode

Comment (yaohaizh) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: After the stack operation (pop/stepin), the UI is out of sync with the underlying JVM debugger status. Thus, send continue event to update the VSCode UI to let the VSCode populate the threading information. The continue event will stop at the first line of the stepin method.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            context.getProtocolServer().sendEvent(new Events.UserNotificationEvent(NotificationType.ERROR, "Current stack frame doesn't support restart."));
+            context.getProtocolServer().sendEvent(new Events.StoppedEvent("restartframe", stackFrameReference.getThread().uniqueID()));
+            return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.RESTARTFRAME_FAILURE, "Failed to restart the selected stack frame.");
+        }
+    }
+
+    private boolean canRestartFrame(IDebugAdapterContext context, StackFrameReference frameReference) {
+        ThreadReference reference = frameReference.getThread();
+        StackFrame[] frames = context.getStackFrameManager().reloadStackFrames(reference);
+        // Current stack frame cannot be the top frame of the thread:
```
Comment (testforstephen) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: // Cannot restart top stack frame.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            StackFrameUtility.pop(frames[0]);
+            --depth;
+        }
+    }
+
+    private void stepin(IDebugAdapterContext context, ThreadReference thread) {
+        StepRequest request = DebugUtility.createStepIntoRequest(thread, context.getStepFilters().classNameFilters);
+        context.getDebugSession().getEventHub().stepEvents().filter(debugEvent -> request.equals(debugEvent.event.request())).take(1).subscribe(debugEvent -> {
+            debugEvent.shouldResume = false;
+            // Have to send to events to keep the UI sync with the step in operations:
```
Comment (testforstephen) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: typos `to` -> `two`.

Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: `with the step in operations` is a little confusing at the quick glance, how about changing it to `with the latest stepIn status`.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    private void popStackFrames(IDebugAdapterContext context, ThreadReference thread, int depth) throws DebugException {
+        while (depth >= 0) {
+            StackFrame[] frames = context.getStackFrameManager().reloadStackFrames(thread);
+            StackFrameUtility.pop(frames[0]);
+            --depth;
+        }
+    }
+
+    private void stepin(IDebugAdapterContext context, ThreadReference thread) {
```
Comment (testforstephen) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: rename to `stepIn`

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @Override
+    public CompletableFuture<Response> handle(Command command, Arguments arguments, Response response, IDebugAdapterContext context) {
+        RestartFrameArguments restartFrameArgs = (RestartFrameArguments) arguments;
+        StackFrameReference stackFrameReference = (StackFrameReference) context.getRecyclableIdPool().getObjectById(restartFrameArgs.frameId);
+
+        if (stackFrameReference == null) {
+            return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.RESTARTFRAME_FAILURE,
+                    String.format("RestartFarme: cannot find the stack frame with frameID %s", restartFrameArgs.frameId));
```
Comment (testforstephen) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: typos `RestartFarme` -> `RestartFrame`

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        for (int i = 0; i <= frameReference.getDepth() + 1; i++) {
+            if (StackFrameUtility.isNative(frames[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void popStackFrames(IDebugAdapterContext context, ThreadReference thread, int depth) throws DebugException {
+        while (depth >= 0) {
```
Comment (testforstephen) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: It seems all stack frames up to the specified depth can be popped by the one invoke. See the doc https://docs.oracle.com/javase/7/docs/jdk/api/jpda/jdi/

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java (hunk 2 above)
Comment (yaohaizh) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: top stack frame is misleading.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java (hunk 5 above)
Comment (yaohaizh) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: Fixed.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java (hunk 4 above)
Comment (yaohaizh) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: Fixed.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java (hunk 6 above)
Comment (yaohaizh) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: Updated.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    private boolean canRestartFrame(IDebugAdapterContext context, StackFrameReference frameReference) {
+        if (!context.getDebugSession().getVM().canPopFrames()) {
+            return false;
+        }
+        ThreadReference reference = frameReference.getThread();
+        StackFrame[] frames = context.getStackFrameManager().reloadStackFrames(reference);
+        // The frame cannot be the first one of the stack:
+        if (frames.length <= frameReference.getDepth() + 1) {
```
Comment (testforstephen) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java]: From the implementation, it looks `the last one`, not `the first one`.

**Pull request merged**